clients/drcachesim/tests/trace_interval_analysis_unit_tests.cpp

```diff
@@ -105,23 +139,177 @@ class test_analyzer_t : public analyzer_t {
     }
 
 private:
     std::unique_ptr<scheduler_t::stream_t> test_stream_;
 };
 
-// Test analysis_tool_t that stores information about when the interval-end
-// events (generate_shard_interval_result and generate_interval_result) were invoked.
+// Dummy analysis_tool_t that does not provide interval results. This helps verify
+// the case when one of the tools run does not implement the interval related APIs.
+class dummy_analysis_tool_t : public analysis_tool_t {
+public:
+    dummy_analysis_tool_t()
+        : generate_snapshot_count_(0)
+    {
+    }
+    bool
+    process_memref(const memref_t &memref) override
+    {
+        return true;
+    }
+    analysis_tool_t::interval_state_snapshot_t *
+    generate_interval_snapshot(uint64_t interval_id) override
+    {
+        ++generate_snapshot_count_;
+        return nullptr;
+    }
+    bool
+    print_results() override
+    {
+        return true;
+    }
+    bool
+    parallel_shard_supported() override
+    {
+        return true;
+    }
+    void *
+    parallel_shard_init(int shard_index, void *worker_data) override
+    {
+        return nullptr;
+    }
+    bool
+    parallel_shard_exit(void *shard_data) override
+    {
+        return true;
+    }
+    bool
+    parallel_shard_memref(void *shard_data, const memref_t &memref) override
+    {
+        return true;
+    }
+    analysis_tool_t::interval_state_snapshot_t *
+    generate_shard_interval_snapshot(void *shard_data, uint64_t interval_id) override
+    {
+        ++generate_snapshot_count_;
+        return nullptr;
+    }
+    analysis_tool_t::interval_state_snapshot_t *
+    combine_interval_snapshot(analysis_tool_t::interval_state_snapshot_t *one,
+                              analysis_tool_t::interval_state_snapshot_t *two) override
+    {
+        FATAL_ERROR("Did not expect combine_interval_snapshot to be invoked");
+        return nullptr;
+    }
+    bool
+    print_interval_results(
+        const std::vector<interval_state_snapshot_t *> &snapshots) override
+    {
+        FATAL_ERROR("Did not expect print_interval_results to be invoked");
+        return true;
+    }
+    bool
+    release_interval_snapshot(
+        analysis_tool_t::interval_state_snapshot_t *snapshot) override
+    {
+        FATAL_ERROR("Did not expect release_interval_snapshot to be invoked");
+        return true;
+    }
+    int
+    get_generate_snapshot_count()
+    {
+        return generate_snapshot_count_;
+    }
+
+private:
+    int generate_snapshot_count_;
+};
+
+// Test analysis_tool_t that records information about when the
+// generate_shard_interval_snapshot and generate_interval_snapshot APIs were invoked.
 class test_analysis_tool_t : public analysis_tool_t {
 public:
-    test_analysis_tool_t()
+    // Describes the point in trace when an interval ends. This is same as the point
+    // when the generate_*interval_snapshot API is invoked.
+    struct interval_end_point_t {
+        memref_tid_t shard_id_;
```

Comment:
style: We're using a trailing underscore for class fields but not for struct fields